Lib/socketserver.py

```diff
@@ -651,24 +652,46 @@ def process_request_thread(self, request, client_address):
         except Exception:
             self.handle_error(request, client_address)
         finally:
-            self.shutdown_request(request)
+            try:
+                self.shutdown_request(request)
+            finally:
+                self._remove_thread()
+
+    def _remove_thread(self):
+        """Remove a current thread from threads list."""
+        thread = threading.current_thread()
+        if not thread.daemon:
+            if self._threads_lock is None:
+                self._threads_lock = threading.Lock()
+            with self._threads_lock:
+                if self._threads is not None:
```

Comment:
I don't think that self._threads can be None there. I suggest to remove the if.

Reply:
If my understanding is correct, self._threads can be "None" here.
It will occur when server_close() called before the process_request_thread reached to this statement.
So my propose is keep this statement or don't set "none" to self._threads on server_close(). for example,
```
def server_close(self):
    super().server_close()
    if self.block_on_close and self._threads_lock is not None:
        with self._threads_lock:
            threads = self._threads
            self._threads = []
        for thread in threads:
            thread.join()
```
(In this case, _remove_thread() will rise ValueError when called server_close() called, but it will be ignored.)
How wold you think this?

Reply:
I agree there is a window where __threads_ is None. Maru's proposed change to _server_close_ looks okay to me. It simplifies the logic a bit too.

Reply:
In b107b913c17f8eab25be0c1c24edfd25bfbed6a9, I've added a test that captures this behavior (where server_close is called before process_request) and in a1aff7673d947fe63e2f303b1766686602c9e4dc, I've addressed the issue.